Mathlib/Tactic/Bound.lean

```diff
@@ -0,0 +1,338 @@
+/-
+Copyright (c) 2024 Geoffrey Irving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving
+-/
+
+import Aesop
+import Mathlib.Analysis.Analytic.Basic
+import Mathlib.Analysis.SpecialFunctions.Trigonometric.Basic
+import Mathlib.Analysis.SpecialFunctions.Log.Basic
+import Mathlib.Analysis.SpecialFunctions.Pow.Real
+import Mathlib.Tactic.Bound.Init
+
+/-!
+## The `bound` tactic
+
+`bound` is an `aesop` wrapper that proves inequalities by straightforward recursion on structure,
+assuming that intermediate terms are nonnegative or positive as needed.  It also has some report
+for guessing where it is unclear where to recurse, such as which side of a `min` or `max` to use
+as the bound or whether to assume a power is less than or greater than one.
+
+The functionality of `bound` overlaps with `positivity` and `gcongr`, but can jump back and forth
+between `0 ≤ x` and `x ≤ y`-type inequalities.  For example, `bound` proves
+  `0 ≤ c → b ≤ a → 0 ≤ a * c - b * c`
+By turning the goal into `b * c ≤ a * c`, then using `mul_le_mul_of_nonneg_right`.  `bound` also
+has used specialized lemmas for goals of the form `1 ≤ x, 1 < x, x ≤ 1, x < 1`.
+
+Additional hypotheses can be passed as `bound [h0, h1 n, ...]`.  This is equivalent to declaring
+them via `have` before calling `bound`.
+
+See `Ray.Tactic.BoundTests` for tests.
+
+### Calc usage
+
+Since `bound` requires the inequality proof to exactly match the structure of the expression, it is
+often useful to iterate between `bound` and `rw / simp` using `calc`.  Here is an example:
+
+```
+-- Calc example: A weak lower bound for `z ← z^2 + c`
+lemma le_sqr_add {c z : ℂ} (cz : abs c ≤ abs z) (z3 : 3 ≤ abs z) :
+    2 * abs z ≤ Complex.abs (z^2 + c) := by
+  calc Complex.abs (z^2 + c)
+    _ ≥ Complex.abs (z^2) - abs c := by bound
+    _ ≥ Complex.abs (z^2) - abs z := by bound
+    _ ≥ (abs z - 1) * abs z := by rw [mul_comm, mul_sub_one, ←pow_two, ←Complex.abs.map_pow]
+    _ ≥ 2 * abs z := by bound
+```
+
+### Aesop heuristics
+
+`bound` uses a `bound` aesop rule set, with rules organized as follows:
+
+#### High-confidence apply rules
+
+1. `norm apply`: Lemmas which close inequalities immediately, such as `le_refl, Real.exp_pos`,
+    triangle inequalities, etc.
+2. `safe 2 apply`: High-confidence lemmas which produce one new goal, such as `inv_le_one`.  Here by
+    high-confidence we mean roughly "are how one would naively prove the inequality, assuming
+    intermediate terms are nonnegative or positive when necessary".  It is important that these are
+    `safe 2`, not `safe 1`, so that `assumption` can fire early in case the naive proof is wrong
+    and we have an assumption that can close the goal immediately.
```

Comment:
I don't follow this (probably because of my lack of experience with `aesop`). What is the connection between `safe 1` and `assumption`?

Reply:
Hmm, this seems like it was me hallucinating, since actually the priority of `Assumption` is -50: https://github.com/leanprover-community/aesop/blob/master/Aesop/BuiltinRules/Assumption.lean#L14.  I'll remove the comment for now, and test whether I can shift things down once I automate priority assignment.